atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CqlExecutorImpl.java

```diff
@@ -63,9 +71,28 @@
             TableReference tableRef,
             List<byte[]> rowsAscending,
             int limit) {
-        String selQuery = "SELECT key, column1, column2 FROM %s WHERE key IN (%s) LIMIT %s;";
-        CqlQuery query = new CqlQuery(selQuery, quotedTableName(tableRef), keys(rowsAscending), limit(limit));
-        return executeAndGetCells(query, rowsAscending.get(0), CqlExecutorImpl::getCellFromRow);
+        String preparedSelQuery = String.format("SELECT key, column1, column2 FROM %s WHERE key = ? LIMIT %d;",
+                quotedTableName(tableRef).getValue(),
+                limit);
+        ByteBuffer queryBytes = ByteBuffer.wrap(preparedSelQuery.getBytes(StandardCharsets.UTF_8));
+        CqlPreparedResult preparedResult = queryExecutor.prepare(queryBytes, Compression.NONE);
```

Comment:
Let's make the prepare and the execute use the same CassandraClient — Cassandra doesn't distribute the prepared statements by default.

If the Cassandra node restarts between the prepare and the execute statements, it will lose all prepared statements. On an extreme scenario, this might happen between the prepare and execute statements. If we tell a Cassandra node to execute a statement it doesn't know about, I think it will throw a `InvalidRequestException` (complete guess — we should probably test this). So we should probably surround the execute statement with try-catch. We should then retry to prepare and to execute.

There're some interesting knowledge about the prepare statements here:
https://docs.datastax.com/en/developer/java-driver/3.1/manual/statements/prepared/#preparing-on-multiple-nodes